### PR TITLE
htlcswitch: batch preimage writes/consistency fix

### PR DESCRIPTION
**Pull request by cfromknecht**
This PR adds batched writes for witnesses discovered in HTLC forwarding. At the same time, we correct a nuanced consistency issue related to a lack of synchronization with the channel state machine. Forcing the individual preimage writes to be synchronized with the link incurs a heavy performance penalty (about 80% from my measurements). Batching these allows us to minimize the number of db transactions required to write the preimages, allowing us to reinsert the batched write into the link's critical path and resolve the possible inconsistency. In fact, the benchmarks actually showed a slight performance improvement, even with the extra write in the critical path.

See commit messages for more details.

### CODE REVIEW

File: htlcswitch/link.go [hunk 1]
```diff
@@ -447,6 +453,22 @@ func (l *channelLink) Stop() {
 
 	close(l.quit)
 	l.wg.Wait()
+
+	// As a final precaution, we will attempt to flush any uncommitted
+	// preimages to the preimage cache. The preimages should be re-delivered
+	// after channel reestablishment, however this adds an extra layer of
+	// protection in case the peer never returns. Without this, we will be
+	// unable to settle any contracts depending on the preimages even though
+	// we had learned them at some point.
```
Comment (joostjager) [inline htlcswitch/link.go]: Just to make sure that I understand this correctly: this is not to miss the opportunity to apply the preimages to the downstream node. But if this was not in, it would not lead to loss of funds would it? We would learn the preimage when the channel is force closed and then settle downstream?

File: htlcswitch/link.go [hunk 2]
```diff
@@ -1473,6 +1489,41 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
...
+		// Since we may have learned new preimages for the first time,
+		// we'll add them to our preimage cache. By doing this, we
+		// ensure any contested contracts watched by any on-chain
+		// arbitrators can now sweep this HTLC on-chain. We delay
+		// committing the preimages until just before accepting the new
+		// remote commitment, as afterwards the peer won't resend the
+		// Settle messages on the next channel reestablishment. Doing so
+		// allows us to more effectively batch this operation, instead
+		// of doing a single write per preimage.
+		if len(l.uncommitedPreimages) > 0 {
```
Comment (joostjager) [inline htlcswitch/link.go]: Is it right that this performance improvement is only observable if multiple htlcs are fulfilled before a new commitsig is sent? So on a very high frequency link?

Comment (cfromknecht) [reply htlcswitch/link.go]: That's correct, yes

File: htlcswitch/link.go (hunk 1 above)
Comment (cfromknecht) [reply htlcswitch/link.go]: assuming the downstream node were to pull on chain, we may learn the preimage then. alternatively, if the peer reconnects and sends the same settle message and the commit sig, that would also suffice.

atm we spawn and forget a goroutine that is meant handle the write of each preimage, which in most cases would probably execute before shutdown/disconnect. however with the addition batching, we can't rely on these background tasks to execute for us, and in fact may not have attempted the write at all. so we add this to ensure that we always attempt to write any received preimages before the link exits, which is akin to what we do now, though the current impl has weaker consistency guarantees

Comment (joostjager) [reply htlcswitch/link.go]: Ok clear. Yes that goroutine was not perfect indeed.

File: channeldb/witness_cache_test.go [hunk 3]
```diff
@@ -107,7 +107,7 @@ func TestWitnessCacheUnknownWitness(t *testing.T) {
 
 	// We'll attempt to add a new, undefined witness type to the database.
 	// We should get an error.
-	err = wCache.AddWitness(234, key[:])
+	err = wCache.AddWitnesses(234, key[:])
 	if err != ErrUnknownWitnessType {
 		t.Fatalf("expected ErrUnknownWitnessType, got %v", err)
 	}
```
Comment (joostjager) [inline channeldb/witness_cache_test.go]: Do you want to test adding an actual batch of preimages?

File: witness_beacon.go [hunk 4]
```diff
@@ -101,16 +101,21 @@ func (p *preimageBeacon) LookupPreimage(payHash []byte) ([]byte, bool) {
...
+	}
+
 	p.Lock()
 	defer p.Unlock()
 
-	srvrLog.Infof("Adding preimage=%x to witness cache", pre[:])
-
 	// First, we'll add the witness to the decaying witness cache.
-	err := p.wCache.AddWitness(channeldb.Sha256HashWitness, pre)
+	err := p.wCache.AddWitnesses(channeldb.Sha256HashWitness, preimages...)
```
Comment (joostjager) [inline witness_beacon.go]: Can this be lifted outside of the lock, as it is happening in a db tx already?

File: contractcourt/channel_arbitrator.go [hunk 5]
```diff
@@ -62,8 +62,13 @@ type WitnessBeacon interface {
 	// True is returned for the second argument if the preimage is found.
 	LookupPreimage(payhash []byte) ([]byte, bool)
 
-	// AddPreImage adds a newly discovered preimage to the global cache.
-	AddPreimage(pre []byte) error
+	// AddPreimages adds a batch of newly discovered preimages to the global
+	// cache, and also signals any subscribers of the newly discovered
+	// witness.
+	//
+	// NOTE: The backing slice of MUST NOT be modified, otherwise the
```
Comment (joostjager) [inline contractcourt/channel_arbitrator.go]: It is a risk. You could use the new `lntypes.Preimage` type and go with copies.

File: htlcswitch/link.go [hunk 6]
```diff
@@ -331,6 +331,12 @@ type channelLink struct {
 	// commitment fee every time it fires.
 	updateFeeTimer *time.Timer
 
+	// uncommitedPreimages stores a list of all preimages that have been
+	// learned since receiving the last CommitSig from the remote peer. The
+	// batch will be flushed just before accepting the subsequent CommitSig
+	// or on shutdown to avoid doing a write for each preimage received.
+	uncommitedPreimages [][]byte
```
Comment (joostjager) [inline htlcswitch/link.go]: Isn't commited spelled as committed?

File: htlcswitch/link.go (hunk 2 above)
Comment (joostjager) [reply htlcswitch/link.go]: Ok. But the main reason for this PR is the unsafety of writing the preimage?

Comment (cfromknecht) [reply htlcswitch/link.go]: The 80% penalty comes when writing each preimage individually, using our async_payments_benchmark integration test as a yard stick (which does attempt to saturate the links). For the single payment case, there would be a negligible penalty since the write is now done in the critical path

File: contractcourt/channel_arbitrator.go (hunk 5 above)
Comment (cfromknecht) [reply contractcourt/channel_arbitrator.go]: Thought about this, but we’d still be forced to perform a deep copy to ensure the notifications can’t be corrupted

Comment (cfromknecht) [reply contractcourt/channel_arbitrator.go]: Nvm I see your point now, using `lntypes.Hash` makes a shallow copy sufficient. Good suggestion, also increases type safety. Will fix!

File: htlcswitch/link.go (hunk 2 above)
Comment (joostjager) [reply htlcswitch/link.go]: Ok, got it.

File: contractcourt/channel_arbitrator.go (hunk 5 above)
Comment (joostjager) [reply contractcourt/channel_arbitrator.go]: yes, but then `lntypes.Preimage` for the preimages

Comment (cfromknecht) [reply contractcourt/channel_arbitrator.go]: haha yes, that's what i meant to type :)

File: channeldb/witness_cache.go [hunk 7]
```diff
@@ -70,12 +70,41 @@ func (d *DB) NewWitnessCache() *WitnessCache {
...
 // TODO(roasbeef): fake closure to map instead a constructor?
-func (w *WitnessCache) AddWitness(wType WitnessType, witness []byte) error {
+func (w *WitnessCache) AddWitnesses(wType WitnessType, witnesses ...[]byte) error {
+	// Define a key-value struct that holds each key -> witness pair.
+	type witnessEntry struct {
+		key     []byte
+		witness []byte
+	}
+
+	// Optimistically compute the witness keys before attempting to start
```
Comment (halseth) [inline channeldb/witness_cache.go]: Awesome change!

File: channeldb/witness_cache_test.go [hunk 8]
```diff
@@ -59,13 +59,13 @@ func TestWitnessCacheDeletion(t *testing.T) {
 	// We'll start by adding two witnesses to the cache.
 	witness1 := rev[:]
 	witness1Key := sha256.Sum256(witness1)
-	if err := wCache.AddWitness(Sha256HashWitness, witness1); err != nil {
+	if err := wCache.AddWitnesses(Sha256HashWitness, witness1); err != nil {
```
Comment (halseth) [inline channeldb/witness_cache_test.go]: Nit: long line

File: contractcourt/channel_arbitrator.go (hunk 5 above)
Comment (halseth) [reply contractcourt/channel_arbitrator.go]: 👍

File: contractcourt/htlc_outgoing_contest_resolver.go [hunk 9]
```diff
@@ -78,7 +78,7 @@ func (h *htlcOutgoingContestResolver) Resolve() (ContractResolver, error) {
 
 		// With the preimage obtained, we can now add it to the global
 		// cache.
-		if err := h.PreimageDB.AddPreimage(preimage[:]); err != nil {
+		if err := h.PreimageDB.AddPreimages(preimage[:]); err != nil {
```
Comment (halseth) [inline contractcourt/htlc_outgoing_contest_resolver.go]: Long line

File: witness_beacon.go (hunk 4 above)
Comment (halseth) [reply witness_beacon.go]: Since we never remove from the cache, I believe that's true

Comment (cfromknecht) [reply witness_beacon.go]: great suggestion!

File: channeldb/witness_cache_test.go (hunk 8 above)
Comment (cfromknecht) [reply channeldb/witness_cache_test.go]: exactly 80

File: contractcourt/htlc_outgoing_contest_resolver.go (hunk 9 above)
Comment (cfromknecht) [reply contractcourt/htlc_outgoing_contest_resolver.go]: i think this is 75 chars

File: channeldb/witness_cache_test.go (hunk 3 above)
Comment (cfromknecht) [reply channeldb/witness_cache_test.go]: indeed! i've also added a test in `link_test.go` to assert the link's behavior is correct

File: contractcourt/channel_arbitrator.go (hunk 5 above)
Comment (cfromknecht) [reply contractcourt/channel_arbitrator.go]: I generally dig the move to use a `Preimage` type for type safety. Unfortunately, it turns out that the `channeldb.WitnessCache` is written in a general sense, meaning that it only operates on byte slices. So even though we get type safety going into the `preimageBeacon`, we then have to make a second copy of all of the preimages before passing them into the `WitnessCache`, so it ends up being pretty ugly.

One thing I've considered is making a wrapper/specialized interface, maybe `AddSha256Preimages` which uses the proper types, then these are serialized into the witness cache directly and hopefully avoiding the redundant copying. Lmk if I should investigate this further

Comment (cfromknecht) [reply contractcourt/channel_arbitrator.go]: ended up going with the a helper that does only sha256 preimages, makes the code much cleaner

**Comment by cfromknecht** [comment]
@joostjager @halseth ready for another round of review. the changes sprawled quite a bit as you'll see. I've also moved the lntypes 

I plan to consolidate 696a5f9 (the move to lntypes) and d47aa70 (first commit), but wanted to get your feedback on the additional commits before doing so.

Along the way:
 - `lnmock` is born, avoiding the same mockPreimageCache implementation in 4 places in the code base.
 - `lntypes` is moved to `lnhash` 
 - 'channeldb.WitnessCache` gained the `AddSHA256Witnesses` helper, to avoid the extra copied in the `preimageBeacon`
 - modfiied the `channeldb/witness_cache_test` to test batch insertion
 - added a link-level test to assert that the changes in `htlcswitch/link.go` behave as expected.

ptal

File: htlcswitch/link_test.go [hunk 10]
```diff
@@ -4500,6 +4634,170 @@ func TestChannelLinkWaitForRevocation(t *testing.T) {
...
+	// committed. On the second, we'll disconnect before allowing Alice to
+	// process the CommitSig, and ensure that the link writes the preimage
+	// batch before exiting.
+	var (
+		start      = 0
+		end        = batchSize
+		disconnect = false
+	)
+
+checkBatchConsistency:
```
Comment (joostjager) [inline htlcswitch/link_test.go]: I think it would be more understandable to use one or two functions instead of the `goto`. Or alternatively make it a parameterized test (with and without disconnect) that is ran from a parent test using `t.Run`.

File: lntypes/preimage.go [hunk 11]
```diff
@@ -9,28 +9,40 @@ import (
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &preimage, nil
+}
+
+// MakePreimage returns a new Preimage from a bytes slice. An error is returned
+// if the number of bytes passed in is not PreimageSize.
+func MakePreimage(newPreimage []byte) (Preimage, error) {
```
Comment (joostjager) [inline lntypes/preimage.go]: Yes I also ran into that ref return from `NewPreimage`. In the hodl invoice branch, I chose to just change the return type to a value. I couldn't see a good use case for the ptr return.

File: lnwallet/channel.go [hunk 12]
```diff
@@ -5394,7 +5395,7 @@ func newOutgoingHtlcResolution(signer input.Signer, localChanCfg *channeldb.Chan
 func newIncomingHtlcResolution(signer input.Signer, localChanCfg *channeldb.ChannelConfig,
 	commitHash chainhash.Hash, htlc *channeldb.HTLC, keyRing *CommitmentKeyRing,
 	feePerKw SatPerKWeight, dustLimit btcutil.Amount, csvDelay uint32,
-	localCommit bool, preimage [32]byte) (*IncomingHtlcResolution, error) {
+	localCommit bool) (*IncomingHtlcResolution, error) {
```
Comment (joostjager) [inline lnwallet/channel.go]: Separate commit as it is more than changing types?

File: lnmock/preimage_cache.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
+package lnmock
```
Comment (joostjager) [inline lnmock/preimage_cache.go]: I am not sure about a mock package. Whether it will again introduce circular dep issues. In general I'd think it is best to define the mock in the same package as the real implementation, so that there can't be new circular dependencies.

File: channeldb/invoices.go [hunk 14]
```diff
@@ -11,7 +11,7 @@ import (
 
 	"github.com/btcsuite/btcd/wire"
 	"github.com/coreos/bbolt"
-	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnhash"
```
Comment (joostjager) [inline channeldb/invoices.go]: `lnhash` is what I had initially, but after comments I changed it to `lntypes`.

File: channeldb/witness_cache.go [hunk 15]
```diff
@@ -63,25 +64,44 @@ type WitnessCache struct {
...
 func (d *DB) NewWitnessCache() *WitnessCache {
 	return &WitnessCache{
 		db: d,
 	}
 }
 
+// AddSHA256Witnesses adds a batch of new SHA256 preimages into the witness
+// cache. This is an alias for AddWitnesses that uses Sha256HashWitness as the
+// preimages' witness type.
+func (w *WitnessCache) AddSHA256Witnesses(preimages ...lnhash.Preimage) error {
```
Comment (joostjager) [inline channeldb/witness_cache.go]: I could only find a single witness type. Why don't we remove it alltogether? Looks like something was added prematurely that we may never need?

File: witness_beacon.go [hunk 16]
```diff
@@ -111,22 +111,15 @@ func (p *preimageBeacon) LookupPreimage(
...
 func (p *preimageBeacon) AddPreimages(preimages ...lnhash.Preimage) error {
 	// Copy the preimages to ensure the backing area can't be modified by
-	// the caller when delivering notifications. We also copy the keys so
-	// that the they can be passed in as a slice of byte slices to the
-	// generic witness cache.
-	preimageCopies := make([]lnhash.Preimage, len(preimages))
-	preimageKeys := make([][]byte, len(preimages))
-	for i, preimage := range preimages {
+	// the caller when delivering notifications.
+	preimageCopies := make([]lnhash.Preimage, 0, len(preimages))
```
Comment (joostjager) [inline witness_beacon.go]: Would reorder commits to avoid adding and removing code again in the same PR

File: channeldb/witness_cache.go (hunk 15 above)
Comment (cfromknecht) [reply channeldb/witness_cache.go]: remove the older interface you mean? we do need to keep the constants to access the witnesses on disk

Comment (joostjager) [reply channeldb/witness_cache.go]: Yes, remove the old interface.

File: lnmock/preimage_cache.go (hunk 13 above)
Comment (cfromknecht) [reply lnmock/preimage_cache.go]: all of the circular dependency issues wrt to mocks are solved by writing unit tests in `package name_test`. there can't be circular dependencies because `name_test` can't be imported by anything else, and the mocks aren't to be used in the real implementation.

if we add them in the same package but make them private, they'll be inaccessible from the tests that do so. if we make them public to access from tests, then we have to document them to satisfy the linter but they also show up on godocs.

File: lnwallet/channel.go (hunk 12 above)
Comment (cfromknecht) [reply lnwallet/channel.go]: hmm must have been a rebase issue, will address when i restructure the commits

File: lntypes/preimage.go (hunk 11 above)
Comment (cfromknecht) [reply lntypes/preimage.go]: sounds good, i'm okay with either approach. indeed i found i didn't need the pointer object in this PR, but i could imagine cases where it would be useful

File: htlcswitch/link_test.go (hunk 10 above)
Comment (cfromknecht) [reply htlcswitch/link_test.go]: good suggestion, i'm actually realizing now that the persistent link harness is unnecessary for this test, so splitting it into a table driven test would be easier now

File: lnmock/preimage_cache.go (hunk 13 above)
Comment (joostjager) [reply lnmock/preimage_cache.go]: Ok. Yes, in another project I actually do the same. Created a separate package containing test code to be able to reuse it without exposing it all from the real implementation.

File: lntypes/preimage.go (hunk 11 above)
Comment (joostjager) [reply lntypes/preimage.go]: I think just one constructor is enough, but it is all not very important.

File: channeldb/witness_cache.go (hunk 15 above)
Comment (cfromknecht) [reply channeldb/witness_cache.go]: Just remembered that we will have other witness types when we move to discrete log  challenges, though seems like that can also be exposed only via the targeted interfaces instead of the general one. Also allows us to make `WitnessType` private

File: channeldb/invoices.go (hunk 14 above)
Comment (cfromknecht) [reply channeldb/invoices.go]: removed import move

File: channeldb/witness_cache.go (hunk 15 above)
Comment (cfromknecht) [reply channeldb/witness_cache.go]: decided to leave the existing interface for now, though i'm not opposed re: removing it in the future. main reason is to keep the test asserting the two are identical, as i did discover a bug in the initial implementation as a result of writing the test

**Comment by cfromknecht** [comment]
@joostjager @halseth comments addressed and commits restructured, ptal

File: contractcourt/htlc_success_resolver.go [hunk 17]
```diff
@@ -103,6 +103,18 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
...
 				h.payHash[:])
 
+			// The wallet returns resolutions with the zero
+			// preimage, as it doesn't have access to the preimage
+			// cache. Before constructing the input, we lookup the
+			// cache and apply it to our resolution.
+			preimage, ok := h.PreimageDB.LookupPreimage(h.payHash)
+			if !ok {
+				err := fmt.Errorf("unable to find preimage "+
+					"for hash=%x", h.payHash)
```
Comment (joostjager) [inline contractcourt/htlc_success_resolver.go]: These formatting string can now also be `%v`. End result is the same though.

File: contractcourt/htlc_incoming_contest_resolver.go [hunk 18]
```diff
@@ -240,6 +215,15 @@ func (h *htlcIncomingContestResolver) AttachResolverKit(r ResolverKit) {
 	h.ResolverKit = r
 }
 
+// applyPreimage sets the preimage for the internal htlc resolution and logs the
+// event.
+func (h *htlcIncomingContestResolver) applyPreimage(preimage lntypes.Preimage) {
```
Comment (joostjager) [inline contractcourt/htlc_incoming_contest_resolver.go]: Instead of this function, can the inner `htlcSuccessResolver.applyPreimage` be called?

File: lnwallet/channel.go [hunk 19]
```diff
@@ -5490,7 +5480,10 @@ func newIncomingHtlcResolution(signer input.Signer, localChanCfg *channeldb.Chan
 	}
 
 	// Next, we'll construct the full witness needed to satisfy the input
-	// of the success transaction.
+	// of the success transaction. We'll populate the witness stack with a
+	// blank preimage, which must be overwritten using ApplyPreimage with
+	// the discovered preimage before the witness can be valid.
+	var preimage [32]byte
 	successWitness, err := input.ReceiverHtlcSpendRedeem(
 		htlc.Signature, preimage[:], signer, &successSignDesc, successTx,
```
Comment (joostjager) [inline lnwallet/channel.go]: Maybe not for this PR, but could eventually this whole witness generation be moved to the resolvers? Not to create a template here and fill in the preimage later?

File: channeldb/witness_cache.go [hunk 20]
```diff
@@ -70,12 +71,69 @@ func (d *DB) NewWitnessCache() *WitnessCache {
...
+	return w.addWitnessEntries(Sha256HashWitness, entries)
+}
+
+// AddWitnesses adds a batch of new witnesses of wType to the witness cache. The
+// type of the witness will be used to map each witness to the key that will be
+// used to look it up. All witnesses should be of the same WitnessType.
 //
 // TODO(roasbeef): fake closure to map instead a constructor?
-func (w *WitnessCache) AddWitness(wType WitnessType, witness []byte) error {
+func (w *WitnessCache) AddWitnesses(wType WitnessType, witnesses ...[]byte) error {
```
Comment (joostjager) [inline channeldb/witness_cache.go]: >decided to leave the existing interface for now, though i'm not opposed re: removing it in the future. main reason is to keep the test asserting the two are identical, as i did discover a bug in the initial implementation as a result of writing the test

If this is dead code now and just needed for testing, I'd suggest moving it to the test.

File: contractcourt/htlc_incoming_contest_resolver.go (hunk 18 above)
Comment (cfromknecht) [reply contractcourt/htlc_incoming_contest_resolver.go]: i wanted to keep the log message that properly displays the resolver type via `%T`, but yes it is possible

File: channeldb/witness_cache.go (hunk 20 above)
Comment (cfromknecht) [reply channeldb/witness_cache.go]: good idea, will fix

File: lnwallet/channel.go (hunk 19 above)
Comment (cfromknecht) [reply lnwallet/channel.go]: indeed, i'm not sure how best to accomplish that atm, but i added a todo

Comment (cfromknecht) [reply lnwallet/channel.go]: added todo

File: lnwallet/channel.go [hunk 21]
```diff
@@ -5194,6 +5194,23 @@ type IncomingHtlcResolution struct {
...
+
+	// If this our commitment transaction, then we'll need to populate the
+	// witness for the second-level HTLC transaction.
+	if r.SignedSuccessTx != nil {
+		// Within the witness for the success transaction, the preimage
+		// is the 4th element as it looks like:
+		//
+		//  * <sender sig> <recvr sig> <preimage> <witness script>
+		r.SignedSuccessTx.TxIn[0].Witness[3] = preimage[:]
+	}
```
Comment (cfromknecht) [inline lnwallet/channel.go]: is there anything we need to set in this case? @halseth @joostjager @Roasbeef

**Comment by cfromknecht** [comment]
@joostjager revamped the witness cache to only expose the specific interface, makes everything much cleaner imo, ptal

**Comment by cfromknecht** [comment]
note: planning to squash `channeldb+witness_beacon: use sha256 lookup+delete witness` down into `channeldb/witness_cache: create AddSHA256Witnesses helper + test`

File: contractcourt/htlc_outgoing_contest_resolver.go [hunk 22]
```diff
@@ -58,38 +60,47 @@ func (h *htlcOutgoingContestResolver) Resolve() (ContractResolver, error) {
...
-			copy(preimage[:], spendingInput.Witness[1])
+			preimageBytes = spendingInput.Witness[1]
+		}
+
+		preimage, err := lntypes.MakePreimage(preimageBytes)
+		if err != nil {
+			return nil, err
 		}
 
 		log.Infof("%T(%v): extracting preimage=%x from on-chain "+
```
Comment (joostjager) [inline contractcourt/htlc_outgoing_contest_resolver.go]: preimage can have %v now that it uses the strong type.

File: contractcourt/htlc_success_resolver.go [hunk 23]
```diff
@@ -191,6 +205,17 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
...
 
+	// The wallet returns resolutions with the zero preimage, as it doesn't
+	// have access to the preimage cache. Before constructing the input, we
+	// lookup the cache and apply it to our resolution.
+	preimage, ok := h.PreimageDB.LookupPreimage(h.payHash)
+	if !ok {
+		err := fmt.Errorf("unable to find preimage "+
+			"for hash=%v", h.payHash)
+		return nil, err
+	}
```
Comment (joostjager) [inline contractcourt/htlc_success_resolver.go]: You could move the lookup code into `applyPreimage` too and rename the method. Prevent duplication from above. But not critical.

File: htlcswitch/link_test.go [hunk 24]
```diff
@@ -4500,6 +4628,135 @@ func TestChannelLinkWaitForRevocation(t *testing.T) {
...
+			name:       "flush on disconnect",
+			disconnect: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			testChannelLinkBatchPreimageWrite(t, test.disconnect)
+		})
+	}
```
Comment (joostjager) [inline htlcswitch/link_test.go]: If there are only two cases (or even a few more), I tend to unroll the loop and just make that calls to `t.Run` one by one. Make it a little lighter. But it is a personal pref.

```
t.Run("a", func(t *testing.T) {test(t, false)}
t.Run("b", func(t *testing.T) {test(t, true)}
```

File: lntypes/preimage.go (hunk 11 above)
Comment (halseth) [reply lntypes/preimage.go]: Would prefer sticking to one constructor to avoid confusion :)

File: contractcourt/channel_arbitrator.go [hunk 25]
```diff
@@ -1209,13 +1207,12 @@ func (c *ChannelArbitrator) checkChainActions(height uint32,
...
-		payHash := htlc.RHash
-
 		// If we have the pre-image, then we should go on-chain to
 		// redeem the HTLC immediately.
-		if _, ok := c.cfg.PreimageDB.LookupPreimage(payHash[:]); ok {
+		if _, ok := c.cfg.PreimageDB.LookupPreimage(htlc.RHash); ok {
 			log.Tracef("ChannelArbitrator(%v): preimage for "+
-				"htlc=%x is known!", c.cfg.ChanPoint, payHash[:])
+				"htlc=%x is known!", c.cfg.ChanPoint,
+				htlc.RHash[:])
```
Comment (halseth) [inline contractcourt/channel_arbitrator.go]: I think the `Hash` type has a `String` method, so you can use "%v" to print

File: contractcourt/htlc_outgoing_contest_resolver.go [hunk 26]
```diff
@@ -2,9 +2,11 @@ package contractcourt
 
 import (
 	"fmt"
-	"github.com/lightningnetwork/lnd/input"
 	"io"
 
+	"github.com/lightningnetwork/lnd/input"
```
Comment (halseth) [inline contractcourt/htlc_outgoing_contest_resolver.go]: nit: imports should be moved by other `lnd` imports

File: contractcourt/htlc_success_resolver.go [hunk 27]
```diff
@@ -3,9 +3,11 @@ package contractcourt
 import (
 	"encoding/binary"
 	"fmt"
-	"github.com/lightningnetwork/lnd/input"
 	"io"
 
+	"github.com/lightningnetwork/lnd/input"
```
Comment (halseth) [inline contractcourt/htlc_success_resolver.go]: same wrt imports

File: witness_beacon.go [hunk 28]
```diff
@@ -66,53 +66,60 @@ func (p *preimageBeacon) SubscribeUpdates() *contractcourt.WitnessSubscription {
...
 	if err != nil {
-		ltndLog.Errorf("unable to lookup witness: %v", err)
-		return nil, false
+		ltndLog.Errorf("Unable to lookup witness: %v", err)
+		return lntypes.Preimage{}, false
+	}
+
+	preimage, err := lntypes.MakePreimage(preimageBytes)
+	if err != nil {
+		ltndLog.Errorf("Unable to build preiamge: %v", err)
```
Comment (halseth) [inline witness_beacon.go]: nit:  typo "preiamge"

File: htlcswitch/link.go [hunk 29]
```diff
@@ -1475,6 +1491,40 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
...
+		// we'll add them to our preimage cache. By doing this, we
+		// ensure any contested contracts watched by any on-chain
+		// arbitrators can now sweep this HTLC on-chain. We delay
+		// committing the preimages until just before accepting the new
+		// remote commitment, as afterwards the peer won't resend the
+		// Settle messages on the next channel reestablishment. Doing so
+		// allows us to more effectively batch this operation, instead
+		// of doing a single write per preimage.
+		if len(l.uncommittedPreimages) > 0 {
+			err := l.cfg.PreimageCache.AddPreimages(
```
Comment (halseth) [inline htlcswitch/link.go]: Does the `AddPreimages` method handle zero-length slices? If it did, we can avoid some indentation here...

File: lnmock/preimage_cache.go (hunk 13 above)
Comment (halseth) [reply lnmock/preimage_cache.go]: I'm not sure about this either. Maybe the `preimage_cache` should have its own package, which can contain both the real, and the mock implementation?

Either way, I think this discussion can be made its own PR, as the actual consistency fix is what is important here.

File: contractcourt/htlc_success_resolver.go [hunk 30]
```diff
@@ -103,6 +103,18 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
 				"incoming+remote htlc confirmed", h,
 				h.payHash[:])
 
+			// The wallet returns resolutions with the zero
+			// preimage, as it doesn't have access to the preimage
+			// cache. Before constructing the input, we lookup the
+			// cache and apply it to our resolution.
+			preimage, ok := h.PreimageDB.LookupPreimage(h.payHash)
```
Comment (halseth) [inline contractcourt/htlc_success_resolver.go]: I'm confused... is this another, unrelated fix to the consistency write fix?

File: lnwallet/channel.go [hunk 31]
```diff
@@ -5591,19 +5584,12 @@ func extractHtlcResolutions(feePerKw SatPerKWeight, ourCommit bool,
...
-			// populate this resolution.
-			preimage, _ := pCache.LookupPreimage(htlc.RHash)
-
-			// Otherwise, we'll create an incoming HTLC resolution
-			// as we can satisfy the contract.
-			var pre [32]byte
-			copy(pre[:], preimage[:])
+			// We'll create an incoming HTLC resolution which allows
+			// us to satisfy the contract assuming the preimage is
+			// discovered before the timelock expires.
```
Comment (halseth) [inline lnwallet/channel.go]: Am I wrong in thinking that the write consistency fix is unrelated to this change?

I'm a bit uneasy doing critical changes like this in a big PR. I would prefer having this in a separate PR where the proper discussion can be had.

File: channeldb/witness_cache.go [hunk 32]
```diff
@@ -78,10 +77,10 @@ type witnessEntry struct {
 	witness []byte
 }
 
-// AddSHA256Witnesses adds a batch of new SHA256 preimages into the witness
+// AddSha256Witnesses adds a batch of new sha256 preimages into the witness
 // cache. This is an alias for AddWitnesses that uses Sha256HashWitness as the
 // preimages' witness type.
-func (w *WitnessCache) AddSHA256Witnesses(preimages ...lntypes.Preimage) error {
+func (w *WitnessCache) AddSha256Witnesses(preimages ...lntypes.Preimage) error {
```
Comment (halseth) [inline channeldb/witness_cache.go]: 👍 I was a bit confused by the introduction of the two methods when reading this commit by commit, so would be nice to squash this with the first change :)

File: contractcourt/htlc_incoming_contest_resolver.go [hunk 33]
```diff
@@ -218,8 +218,8 @@ func (h *htlcIncomingContestResolver) AttachResolverKit(r ResolverKit) {
 // applyPreimage sets the preimage for the internal htlc resolution and logs the
 // event.
 func (h *htlcIncomingContestResolver) applyPreimage(preimage lntypes.Preimage) {
-	log.Infof("%T(%v): extracted preimage=%x from beacon!", h,
-		h.htlcResolution.ClaimOutpoint, preimage[:])
+	log.Infof("%T(%v): extracted preimage=%v from beacon!", h,
```
Comment (halseth) [inline contractcourt/htlc_incoming_contest_resolver.go]: 👍

File: channeldb/witness_cache.go (hunk 32 above)
Comment (halseth) [reply channeldb/witness_cache.go]: Ok, I see now that this is planned 👍

File: lntypes/preimage.go (hunk 11 above)
Comment (joostjager) [reply lntypes/preimage.go]: In any case, please keep the non-pointer named `Make`. In another PR I just went through a rebase action to do that. I understand that there is a convention where `New` is for pointers and `Make` for values.

File: htlcswitch/link.go (hunk 29 above)
Comment (cfromknecht) [reply htlcswitch/link.go]: good suggestion, added a check in `AddPreimages` to exit early for zero length slices

File: lntypes/preimage.go (hunk 11 above)
Comment (cfromknecht) [reply lntypes/preimage.go]: My personal preference is typically to add both constructors, which allows the caller to use whatever is best as requiring an extra indirect or dereference can have performance impacts or make calling code more verbose. It seems like we don't really use the pointer version tho atm, so i just decided to leave `MakePreimage` that returns the non-pointer version. We can always add `NewPreimage` later if we want/need to

File: contractcourt/channel_arbitrator.go (hunk 25 above)
Comment (cfromknecht) [reply contractcourt/channel_arbitrator.go]: `htlc.RHash` is still a `[32]byte`, still need `%x`

**Comment by cfromknecht** [comment]
@halseth @joostjager i took out the commits that removed the preimage cache from lnwallet, should be a little more contained now. it still includes the commits that make the switch to `lntypes` as requested, ptal

File: channeldb/witness_cache.go [hunk 34]
```diff
@@ -77,6 +78,24 @@ type witnessEntry struct {
...
 
+// AddSha256Witnesses adds a batch of new sha256 preimages into the witness
+// cache. This is an alias for AddWitnesses that uses Sha256HashWitness as the
+// preimages' witness type.
+func (w *WitnessCache) AddSha256Witnesses(preimages ...lntypes.Preimage) error {
+	// Optimistically compute the preimages' hashes before attempting to
+	// start the db transaction.
+	entries := make([]witnessEntry, 0, len(preimages))
+	for i := range preimages {
+		hash := preimages[i].Hash()
```
Comment (halseth) [inline channeldb/witness_cache.go]: any reason not to use `for _, p := range preimages {`?

File: contractcourt/channel_arbitrator.go (hunk 25 above)
Comment (halseth) [reply contractcourt/channel_arbitrator.go]: Ah, but it can still be used when calling `LookupPreimage(payhash lntypes.Hash)`? Did not know that :)

File: channeldb/witness_cache.go (hunk 34 above)
Comment (cfromknecht) [reply channeldb/witness_cache.go]: Indeed because each `p[:]` will point to the same preimage due to range var binding. This was the bug I had initially ;)

File: contractcourt/channel_arbitrator.go (hunk 25 above)
Comment (cfromknecht) [reply contractcourt/channel_arbitrator.go]: yep, since `lntypes.Preimage` is a named type and `[32]byte` is an unnamed type, they can be casted implicitly bc the underlying raw type is the same

File: contractcourt/htlc_success_resolver.go (hunk 23 above)
Comment (cfromknecht) [reply contractcourt/htlc_success_resolver.go]: good suggestion, will carry this over to a subsequent PR that removes the preimage cache from lnwallet

File: htlcswitch/link_test.go (hunk 24 above)
Comment (cfromknecht) [reply htlcswitch/link_test.go]: ended up leaving as is so that the the booleans are named :)
